dgl_sparse/src/spspmm.cc

```diff
@@ -116,10 +116,44 @@ tensor_list SpSpMMAutoGrad::backward(
   return {torch::Tensor(), lhs_val_grad, torch::Tensor(), rhs_val_grad};
 }
 
+c10::intrusive_ptr<SparseMatrix> DiagSpSpMM(
+    const c10::intrusive_ptr<SparseMatrix>& lhs_mat,
+    const c10::intrusive_ptr<SparseMatrix>& rhs_mat) {
+  if (lhs_mat->HasDiag()) {
+    if (rhs_mat->HasDiag()) {
+      // Diag @ Diag
+      const int64_t m = lhs_mat->shape()[0];
+      const int64_t n = lhs_mat->shape()[1];
+      const int64_t p = rhs_mat->shape()[1];
+      const int64_t common_diag_len = std::min({m, n, p});
+      const int64_t new_diag_len = std::min(m, p);
+      auto slice = torch::indexing::Slice(0, common_diag_len);
+      auto new_val =
+          lhs_mat->value().index({slice}) * rhs_mat->value().index({slice});
+      new_val = torch::constant_pad_nd(
+          new_val, {0, new_diag_len - common_diag_len}, 0);
+      return SparseMatrix::FromDiag(new_val, {m, p});
+    } else {
+      // Diag @ Sparse
+      auto row = rhs_mat->Indices().index({0});
+      auto val = lhs_mat->value().index_select(0, row) * rhs_mat->value();
+      return SparseMatrix::ValLike(rhs_mat, val);
```

Comment:
Is this implementation correct? The index_select should use `A.row` according to https://github.com/dmlc/dgl/blob/master/python/dgl/sparse/matmul.py#L172.

Reply:
I think so. `rhs_mat->Indices().index({0})` returns the row coordinates of the SparseMatrix without data copy.

Reply:
I see. `rhs_mat->Indices()` returns the COO index. How is it different with `rhs_mat->COOPointer()->indices`? Do we want to have two interfaces?

Reply:
They are the same. But we need `Indices()` exposed to python to avoid data copy of the COO tensor.